### ISSUE TITLE
Clearing harvest source not completing

### ISSUE THREAD

**Issue by chrism80026** [ckan/ckanext-harvest#273]
I have several harvest jobs that were left in limbo in a running state. My understanding is that the right way to fix this to clear the harvest source, but running clear for a harvest source isn't completing for large sources with a lot of past jobs. One example is a harvest source that has 29K records and 72 harvest jobs. Clear ran for 4 days and never finished. From what I can determine, deleting from the resource table is what isn't finishing. Doing a select count on the resource table, I see there are1.4M rows that need to be deleted from resource. That's a big number, but shouldn't take 4 days.  Is there some way to fix the running state of a harvest job? Can I change harvest_job.status to "Finished" to get it to harvest again?  Also is there a way to prune harvest job history short of clearing everything? I'm currently running CKAN 2.1a but moving to 2.3.5 soon.

Here is the delete query from ckanext-harvest/ckanext/harvest/logic/action/update.py that's not completing:

_**delete from resource where resource_group_id in (select id from resource_group where package_id in (select id from package where state = 'to_delete'));**_

**Comment by davidread** [comment]
It's possible the clear command has got into a database deadlock. It may help to stop CKAN/apache running while you do the clear to avoid this. You could try opening 'psql ckan' and playing about with that sql command yourself.

Even better though, the 'job_abort' command was added relatively recently to get out of limbo states:

    paster harvester job_abort {source-id/name}

**Comment by chrism80026** [comment]
Thanks! I upgraded from version .2 to 0.0.5.  I can now abort jobs. Things seem to be running correctly, but I'm now having issues with the CKAN API. This query  (https://data.noaa.gov/api/3/action/package_search?q=type:harvest&rows=1000) returns 0 results. Before the upgrade it returned all the harvest sources. I tried doing an uninstall before installing the newer version but that put the environment in a bad state so I did pip install --upgrade. Any suggestions on fixing this? We currently running CKAN 2.1a.

**Comment by davidread** [comment]
Firstly, avoid "pip install --upgrade" because that upgrades all the dependencies to their latest versions, rather than what was previously installed with a requirements.txt. I suggest you now to `pip install -r ../ckanext-harvest/pip-requirements.txt` to be sure.

The version was stuck at .2 for a while, so it's better to refer to git changerefs.

https://data.noaa.gov/api/3/action/package_search?q=type:harvest&rows=1000 returns 92 result for me. I guess you're referring to a separate dev instance where you run the same query? 

I guess the first thing to do is to check the query is still correct. This isn't something I currently use, so perhaps @amercader can pitch in here.

Then I suggest you check the model migration happened ok. See the migration steps here:
https://github.com/ckan/ckanext-harvest/blob/master/ckanext/harvest/model/__init__.py#L72
where it checks if the title and frequency columns exist on the harvest_source table. Check before and after your upgrade to see if the migration is working. You trigger the migration by running any of the ckanext-harvest paster commands. I believe the server needs to be down at the time or it might try to migrate several times in parallel.

**Comment by davidread** [comment]
I'm closing this issue, since the original problem is solved. If you want to discuss the upgrade problem further, please open a separate issue.
